### PR TITLE
Improve annotation types

### PR DESCRIPTION
**Pull request by robertknight**
This PR contains some typing improvements extracted out of https://github.com/hypothesis/client/pull/4193.

It improves the types for various annotation-related parameters in the `AnnotationsService` service (first commit) and the annotations store module (second commit).

The most notable change is the introduction of a `SavedAnnotation` type for use in situations where a method/function expects to be passed an already-saved annotation, which we know will have an ID assigned. There is an associated `isSaved` function. Calls to `isSaved` have been added in various places to explicitly check that an annotation was saved, but this should already have been the case, so there should be no functional changes.

### CODE REVIEW

File: src/sidebar/components/Annotation/Annotation.js [hunk 1]
```diff
@@ -57,7 +58,11 @@ function Annotation({
   const showActions = !isSaving && !isEditing;
   const showReplyToggle = !isReply && !hasAppliedFilter && replyCount > 0;
 
-  const onReply = () => annotationsService.reply(annotation, userid);
+  const onReply = () => {
+    if (annotation && isSaved(annotation)) {
+      annotationsService.reply(annotation, userid);
+    }
+  };
```
Comment (robertknight) [inline src/sidebar/components/Annotation/Annotation.js:65]: This callback is only used if the annotation is not being edited and not being saved, so the added check here is strictly redundant, but it's much easier to see now that `reply` will never be passed an unsaved annotation.

File: src/sidebar/components/Annotation/Annotation.js [hunk 2]
```diff
@@ -111,7 +116,7 @@ function Annotation({
                 Saving...
               </div>
             )}
-            {annotation && showActions && (
+            {annotation && showActions && isSaved(annotation) && (
```
Comment (robertknight) [inline src/sidebar/components/Annotation/Annotation.js:119]: Given the current UI design and workflow, `showActions` implicitly implies that the annotation is saved, but code is now more explicit that the action bar won't be shown for unsaved annotations.

File: src/sidebar/helpers/annotation-metadata.js [hunk 3]
```diff
@@ -123,11 +126,21 @@ export function isReply(annotation) {
...
+ * @return {annotation is SavedAnnotation}
+ */
+export function isSaved(annotation) {
+  return !!annotation.id;
+}
+
+/**
+ * Return true if an annotation has not been saved to the backend.
  *
+ * @deprecated - Use {@link isSaved} instead
```
Comment (robertknight) [inline src/sidebar/helpers/annotation-metadata.js:143]: Various `isNew(annotation)` calls should be replaced with `!isSaved(annotation)`. This function has been left to facilitate gradual migration.

File: src/sidebar/store/modules/annotations.js [hunk 4]
```diff
@@ -117,7 +125,8 @@ const initialState = {
 /** @typedef {typeof initialState} State */
 
 const reducers = {
-  ADD_ANNOTATIONS: function (state, action) {
+  /** @param {State} state */
+  ADD_ANNOTATIONS(state, action) {
```
Comment (robertknight) [inline src/sidebar/store/modules/annotations.js:129]: I haven't gone through and typed all the `action` parameters yet. That will happen in a separate pass.

File: src/sidebar/store/modules/annotations.js [hunk 5]
```diff
@@ -230,9 +246,10 @@ const reducers = {
         });
         if (newAnn.moderation) {
           const countDelta = action.isFlagged ? 1 : -1;
-          newAnn.moderation = Object.assign({}, annot.moderation, {
-            flagCount: annot.moderation.flagCount + countDelta,
-          });
+          newAnn.moderation = {
+            ...newAnn.moderation,
+            flagCount: newAnn.moderation.flagCount + countDelta,
```
Comment (robertknight) [inline src/sidebar/store/modules/annotations.js:251]: In this context `annot.moderation` and `newAnn.moderation` refer to the same object, but due to the `if` check above, TS can see that `newAnn.moderation` is not null.

File: src/sidebar/components/Annotation/Annotation.js (hunk 1 above)
Comment (lyzadanger) [reply src/sidebar/components/Annotation/Annotation.js:65]: I appreciate this as it expresses an expectation that was implicit before.

File: src/sidebar/components/Annotation/Annotation.js (hunk 2 above)
Comment (lyzadanger) [reply src/sidebar/components/Annotation/Annotation.js:119]: Yes, good, I like.

File: src/sidebar/helpers/annotation-metadata.js (hunk 3 above)
Comment (lyzadanger) [reply src/sidebar/helpers/annotation-metadata.js:143]: This is nicely handled. Thanks! I do notice that the amount of usage of `isNew` is relatively minimal; converting to `!isSaved` could be a pretty quick follow-up step.

File: src/sidebar/services/annotations.js [hunk 6]
```diff
@@ -51,7 +52,7 @@ export class AnnotationsService {
   /**
    * Extend new annotation objects with defaults and permissions.
    *
-   * @param {AnnotationData} annotationData
+   * @param {Omit<AnnotationData, '$tag'>} annotationData
```
Comment (lyzadanger) [inline src/sidebar/services/annotations.js:55]: I wonder if this deserves an explanatory comment? Or, alternately, some sort of adjustment of the `annotationData` type to explicitly separate out that `$tag` is assigned by the client during initialization? Or...?

File: src/sidebar/store/modules/annotations.js [hunk 7]
```diff
@@ -255,7 +272,9 @@ const actions = util.actionTypes(reducers);
 function addAnnotations(annotations) {
   return function (dispatch, getState) {
     const added = annotations.filter(annot => {
-      return !findByID(getState().annotations.annotations, annot.id);
+      return (
+        !annot.id || !findByID(getState().annotations.annotations, annot.id)
+      );
```
Comment (lyzadanger) [inline src/sidebar/store/modules/annotations.js:277]: Is this a "just because you were here" change or is it more meaningful?

File: src/sidebar/store/modules/annotations.js [hunk 8]
```diff
@@ -551,12 +576,13 @@ const orphanCount = createSelector(
...
+ * @param {State} state
+ * @return {SavedAnnotation[]}
  */
 function savedAnnotations(state) {
-  return state.annotations.filter(ann => {
-    return !metadata.isNew(ann);
-  });
+  return /** @type {SavedAnnotation[]} */ (
+    state.annotations.filter(ann => isSaved(ann))
+  );
```
Comment (lyzadanger) [inline src/sidebar/store/modules/annotations.js:585]: I'm curious why the extra typing is needed here.

**Review by lyzadanger — APPROVED** [review]
This is quite nice, and helpful. There's long been some unpleasant ambiguity here.

I wonder how we might handle the local annotation properties (`$tag`, `$anchorTimeout`) going forward. There are a few cases here where you explicitly `Omit<>` one or the other and I'm not sure if that is suggestive of more tweaking or...not a thing.

File: src/sidebar/store/modules/annotations.js (hunk 7 above)
Comment (robertknight) [reply src/sidebar/store/modules/annotations.js:277]: This code was previously incorrect, as it did not account for `addAnnotations` being passed an unsaved annotation. As far as I can tell the effect of this mistake would be to not apply the anchoring timeout when a new unsaved annotation was added to the store, but only if there were already other unsaved annotations. This is an edge case that would be very hard to test manually as it would require that an annotation the user has just created to fail to anchor in the same document.

File: src/sidebar/store/modules/annotations.js (hunk 8 above)
Comment (robertknight) [reply src/sidebar/store/modules/annotations.js:585]: The immediate answer was because TS failed to figured out the effect of `filter` automatically. Upon further experimentation, I found that this works without additional type annotations:

```js
return state.annotations.filter(isSaved);
```

The issue seems to be that the `ann => isSaved(ann)` callback loses the nuance of the `isSaved` return type and treats it as just `boolean`.

The problem with this workaround is that if `isSaved` ever gains support for additional parameters, this could have unexpected interactions with `filter`.

**Comment by robertknight** [comment]
> I wonder how we might handle the local annotation properties ($tag, $anchorTimeout) going forward.

The `Omit` case is used to remove properties from a type when that property _may_ not be present, in a particular context. I'm not quite sure what the end result of this will look like - whether we'll be able to initialize all the local properties in one place or not.

File: src/sidebar/services/annotations.js (hunk 6 above)
Comment (robertknight) [reply src/sidebar/services/annotations.js:55]: `AnnotationData` values do have tags in the places where we currently use them. The case where we pass in an `AnnotationData` without a tag is in methods like `createPageNote` and `reply` in this service.
